### ISSUE TITLE
Suggestion: Don't write into parameter array of an unsealed type

### ISSUE THREAD

**Issue by Joe4evr** [dotnet/roslyn-analyzers#5370]
### Describe the problem you are trying to solve

Array covariance has been a thing since .NET 1.0, and [removing it is not an option](https://github.com/dotnet/csharplang/discussions/5010#discussioncomment-1161043), but it does leave a nasty hole in the type system. Newcomers can easily be thrown for a loop because of it.

### Describe suggestions on how to achieve the rule

In the following example, the assignment in method `N` should be flagged, since `animals` is received as a parameter and the array element is not a sealed type.
```cs
void M()
{
    var giraffes = new Giraffe[] { new Giraffe() };
    N(giraffes);
}

void N(Animal[] animals)
{
    animals[0] = new Turtle();
}

public abstract class Animal { }
public abstract class Mammal : Animal { }
public abstract class Reptile : Animal { }
public sealed class Giraffe : Mammal { }
public sealed class Turtle : Reptile { }
```

This issue might even creep in without much fanfare. Consider a third-party package defines `public sealed class Foo` and for whatever reason, you have a method that takes in a `Foo[]` and writes in its own instances, because for sealed types it's A-OK.
Later, the package author realized they need a `class Bar : Foo` and unsealed `Foo` for that reason (because unsealing a sealed type is not a binary breaking change). Upon updating the package, this analyzer would also immediately inform you that `Foo` is now unsealed and you need to adapt accordingly.

### Additional context

I'm not certain there's a general purpose fix for this issue:
* It's not always possible to change the element type to be sealed because it may be defined outside of the developer's control
* Changing to take in one of the invariant collection types/interfaces could break call-sites
* Changing to take in one of the covariant collection interfaces could be fine, since it's the write into the array that's suspect in the first place, buuuuuuut......
* The method signature could be mandated by a base interface/class, or a delegate conversion
* If the parameter happens to be `params`, it must be an array (until `params {RO}Span`\*/`params IEnumerable` become a thing)

But the bottom line is that if a method takes in some array of an unsealed type, that method *should* only perform reads from said array, and this analyzer would guide developers into that direction.

\*(I know [`params Span` is something the C# LDT is actively looking into](https://github.com/dotnet/csharplang/issues/1757), and a semi-related suggestion is an analyzer that informs users to prefer `params ReadOnlySpan<T>` in practically all cases once that feature drops.)

**Comment by svick** [comment]
>  the assignment in method `N` should be flagged, since `animals` is received as a parameter and the array element is not a sealed type

I disagree. Setting a `Turtle` as an item in an `Animal[]` is not the problematic operation (even if it's the operation that throws), it's assigning `Giraffe[]` into a variable of type `Animal[]` that should be flagged. Otherwise, I think this analyzer would have too many false positives.
